File: frontend/src/app/core/services/user.service.ts
import {Injectable} from "@angular/core";
import {HttpClient} from "@angular/common/http";
import {BehaviorSubject, catchError, Observable, Subject} from "rxjs";
import {ApiResponse, PagingResponse, SortBy} from "../types/api-response.type";
import {environment} from "../../../environments/environment";
import {ChangeInformation, ChangePassword, ForgotPassword, User} from "../types/user.type";
import {handleError} from "../../shared/commons/handle-error-http";
import {Notification, RpNotification} from "../types/noti.type";

@Injectable({
    providedIn:"root"
})

export class UserService{

    notificationState$: BehaviorSubject<RpNotification | null> = new BehaviorSubject<RpNotification | null>(null)
    constructor(private http:HttpClient) {
    
    }

    getEmailForgotPassword(email:string):Observable<ApiResponse<boolean>>{
        return this.http.get<ApiResponse<boolean>>(`${environment.apiUrl}/users/${email}/forgot-mail`).pipe(
            catchError(handleError)
        )
    }
    
    forgotPassword(data:ForgotPassword):Observable<ApiResponse<boolean>>{
        return this.http.patch<ApiResponse<boolean>>(`${environment.apiUrl}/users/forgot`,data).pipe(
            catchError(handleError)
        )
    }

    changePassword(data:ChangePassword):Observable<ApiResponse<boolean>>{
        return this.http.patch<ApiResponse<boolean>>(`${environment.apiUrl}/users/password`,data).pipe(
            catchError(handleError)
        )
    }

    changeInformation(data:ChangeInformation):Observable<ApiResponse<boolean>>{
        return this.http.patch<ApiResponse<boolean>>(`${environment.apiUrl}/users/information`,data).pipe(
            catchError(handleError)
        )
    }

    changeAvatar(avatar:string):Observable<ApiResponse<boolean>>{
        return this.http.patch<ApiResponse<boolean>>(`${environment.apiUrl}/users/avatar`,avatar).pipe(
            catchError(handleError)
        )
    }

    getAuthor(userName:string):Observable<ApiResponse<User>>{
        return this.http.get<ApiResponse<User>>(`${environment.apiUrl}/users/${userName}`).pipe(
            catchError(handleError)
        )
    }

    getListFollowing(userName:string,pageIndex:number,sortBy:SortBy):Observable<ApiResponse<PagingResponse<User[]>>>{
        return this.http.get<ApiResponse<PagingResponse<User[]>>>(`${environment.apiUrl}/users/${userName}/follower?pageIndex=${pageIndex}&sortBy=${sortBy}`).pipe(
            catchError(handleError)
        )
    }
    getListFollowers(userName:string,pageIndex:number,sortBy:SortBy):Observable<ApiResponse<PagingResponse<User[]>>>{
        return this.http.get<ApiResponse<PagingResponse<User[]>>>(`${environment.apiUrl}/users/${userName}/following?pageIndex=${pageIndex}&sortBy=${sortBy}`).pipe(
            catchError(handleError)
        )
    }

    getTop10Notification():Observable<ApiResponse<RpNotification>>{
        return this.http.get<ApiResponse<RpNotification>>(`${environment.apiUrl}/users/top10-noti`).pipe(
            catchError(handleError)
        )
    }

    getAllNotification(pageIndex:number):Observable<ApiResponse<PagingResponse<RpNotification>>>{
        return this.http.get<ApiResponse<PagingResponse<RpNotification>>>(`${environment.apiUrl}/users/notifications?pageIndex=${pageIndex}`).pipe(
            catchError(handleError)
        )
    }

    readNotification(id:number):Observable<ApiResponse<boolean>>{
        return this.http.patch<ApiResponse<boolean>>(`${environment.apiUrl}/users/${id}/notifications`,{}).pipe(
            catchError(handleError)
        )
    }

    readAllNotification():Observable<ApiResponse<boolean>>{
        return this.http.patch<ApiResponse<boolean>>(`${environment.apiUrl}/users/notifications`,{}).pipe(
            catchError(handleError)
        )
    }
}

File: backend/src/main/java/com/example/blog_springboot/modules/user/service/UserService.java
package com.example.blog_springboot.modules.user.service;

import com.example.blog_springboot.commons.PagingResponse;
import com.example.blog_springboot.commons.SuccessResponse;
import com.example.blog_springboot.modules.user.dto.ChangeInformationDTO;
import com.example.blog_springboot.modules.user.dto.ChangePasswordDTO;
import com.example.blog_springboot.modules.user.dto.ChangePermissionDTO;
import com.example.blog_springboot.modules.user.dto.ForgotPasswordDTO;
import com.example.blog_springboot.modules.user.model.User;
import com.example.blog_springboot.modules.user.viewmodel.UserDetailVm;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public interface UserService {
    public SuccessResponse<Boolean> changePassword(ChangePasswordDTO dto,User userPrincipal);
    public SuccessResponse<String> changeAvatar(String avatar,User userPrincipal);

    public SuccessResponse<UserDetailVm> changeInformation(ChangeInformationDTO dto, User userPrincipal);

    public SuccessResponse<UserDetailVm> getMe(User user);

    public SuccessResponse<Boolean> changePermission(ChangePermissionDTO dto);

    public SuccessResponse<Boolean> forgotPassword(ForgotPasswordDTO dto);

    public SuccessResponse<Boolean> sendCodeForgotPassword(String email);

    public SuccessResponse<PagingResponse<List<UserDetailVm>>> getListFollowing(String sortBy, int pageIndex, String userName);

    public SuccessResponse<PagingResponse<List<UserDetailVm>>> getListFollowers(String sortBy, int pageIndex, String userName);

    public SuccessResponse<UserDetailVm> getAuthor(String userName);


}
